src/server/server.ts

```diff
@@ -155,6 +155,52 @@ namespace ts.server {
 
     var logger = createLoggerFromEnv();
 
+    var messagesToWrite: string[] = [];
+    function addMessage(message: string) {
+        messagesToWrite.push(message);
+        // If the current message list has more than 1 messages, that means
+        // the current writing is not ended yet, so don't start new writeNext
+        // as it may interfere with ongoing writing sessions.
+        if (messagesToWrite.length === 1) {
+            startWrite();
+        }
+    }
+
+    function startWrite() {
+        if (messagesToWrite.length === 0) {
+            return;
+        }
+
+        let messageToWrite = messagesToWrite[0];
+        let buffer = new Buffer(messageToWrite, "utf8");
+        write(buffer);
+    }
+
+    function writeNext() {
+        if (messagesToWrite.length > 0) {
+            messagesToWrite = copyListRemovingItem(messagesToWrite[0], messagesToWrite);

```

Comment:
why do you need to copy the list (given that `messagesToWrite` is effectively a queue so items are added\removed at known positions)? You can do better with `messagesToWrite.shift`

```ts
    let pending: string[] = [];
    function queueMessage(s: string) {
        pending.push(s);
        if (pending.length === 1) {
            drain();
        }
    }

    function drain() {
        Debug.assert(pending.length > 0);
        writeBuffer(new Buffer(pending[0], "utf8"), 0);
    }

    function writeBuffer(buffer: Buffer, offset: number) {
        const toWrite = buffer.length - offset;
        fs.write(1, buffer, offset, toWrite, undefined, (err, written, pos) => {
            if (toWrite > written) {
                writeBuffer(buffer, offset + written);
            }
            else {
                Debug.assert(pending.length > 0);
                pending.shift();
                if (pending.length > 0) {
                    drain();
                }
            }
        });
    }
```